File: src/test/java/org/panteleyev/passwdgen/GeneratorTest.java
/*
 Copyright (c) Petr Panteleyev. All rights reserved.
 Licensed under the BSD license. See LICENSE file in the project root for full license information.
 */
package org.panteleyev.passwdgen;

import java.util.Arrays;
import java.util.List;
import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;
import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertFalse;

@Test
public class GeneratorTest {
    @DataProvider(name="testPasswordContentData")
    public Object[][] testPasswordContentData() {
        return new Object[][] {
            { true, false, false, false, false },
            { false, true, false, false, false },
            { false, false, true, false, false },
            { false, false, false, true, false },
            { false, false, false, true, true },
            { true, true, true, true, true },
            { true, true, true, true, false },
        };
    }

    @Test
    public void testPasswordLength() {
        var g = new Generator();
        g.useDigitsProperty().set(true);
        g.useLowerCaseProperty().set(true);
        g.useUpperCaseProperty().set(true);
        g.useSymbolsProperty().set(true);

        Arrays.stream(PasswordLength.values()).forEach(length -> {
            g.lengthProperty().set(length);
            g.generate();
            var res = g.passwordProperty().get();
            assertEquals(res.length(), length.getLength());
        });
    }

    private boolean contains(List<Character> array, char c) {
        return array.contains(c);
    }

    @Test(dataProvider="testPasswordContentData")
    public void testPasswordContent(boolean useDigits, boolean useUpperCase, boolean useLowerCase, boolean useSymbols, boolean avoid) {
        var g = new Generator();

        g.useDigitsProperty().set(useDigits);
        g.useLowerCaseProperty().set(useLowerCase);
        g.useUpperCaseProperty().set(useUpperCase);
        g.useSymbolsProperty().set(useSymbols);
        g.lengthProperty().set(PasswordLength.THIRTY_TWO);
        g.avoidAmbiguousLettersProperty().set(avoid);
        g.generate();

        String res = g.passwordProperty().get();

        for (int i = 0; i < res.length(); i++) {
            var ch = res.charAt(i);

            assertFalse(!useDigits && contains(Generator.DIGITS, ch));
            assertFalse(!useUpperCase && contains(Generator.UPPER_CASE_CHARS, ch));
            assertFalse(!useLowerCase && contains(Generator.LOWER_CASE_CHARS, ch));
            assertFalse(!useSymbols && contains(Generator.SYMBOLS, ch));
            assertFalse(avoid && contains(Generator.BAD_LETTERS, ch));
        }
    }

    @Test
    public void testNoCharacterSetSelected() throws Exception {
        var g = new Generator();
        g.useDigitsProperty().set(false);
        g.useLowerCaseProperty().set(false);
        g.useUpperCaseProperty().set(false);
        g.useSymbolsProperty().set(false);
        g.lengthProperty().set(PasswordLength.FOUR);
        g.generate();
        assertEquals(g.passwordProperty().get(), Generator.ERROR_MESSAGE);
    }
}

File: src/test/java/org/panteleyev/passwdgen/GeneratorTest.kt
/*
 Copyright © 2021-2023 Petr Panteleyev <petr@panteleyev.org>
 SPDX-License-Identifier: BSD-2-Clause
 */
package org.panteleyev.passwdgen

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

class GeneratorTest {
    companion object {
        @JvmStatic
        fun testPasswordContentData(): List<Arguments> {
            return listOf(
                Arguments.of(true, false, false, false, false),
                Arguments.of(false, true, false, false, false),
                Arguments.of(false, false, true, false, false),
                Arguments.of(false, false, false, true, false),
                Arguments.of(false, false, false, true, true),
                Arguments.of(true, true, true, true, true),
                Arguments.of(true, true, true, true, false)
            )
        }
    }

    @Test
    fun testPasswordLength() {
        PasswordLength.values().forEach {
            assertEquals(it.length, Generator.generate(length = it).length)
        }
    }

    private fun contains(array: List<Char>, c: Char): Boolean {
        return array.contains(c)
    }

    @ParameterizedTest
    @MethodSource("testPasswordContentData")
    fun testPasswordContent(
        useDigits: Boolean,
        useUpperCase: Boolean,
        useLowerCase: Boolean,
        useSymbols: Boolean,
        avoid: Boolean
    ) {
        for (ch in Generator.generate(
            upperCase = useUpperCase,
            lowerCase = useLowerCase,
            digits = useDigits,
            symbols = useSymbols,
            noAmbiguousLetters = avoid
        )) {
            assertFalse(!useDigits && contains(Generator.DIGITS, ch))
            assertFalse(!useUpperCase && contains(Generator.UPPER_CASE_CHARS, ch))
            assertFalse(!useLowerCase && contains(Generator.LOWER_CASE_CHARS, ch))
            assertFalse(!useSymbols && contains(Generator.SYMBOLS, ch))
            assertFalse(avoid && contains(Generator.BAD_LETTERS, ch))
        }
    }

    @Test
    fun testNoCharacterSetSelected() {
        assertEquals(Generator.ERROR_MESSAGE, Generator.generate(
            upperCase = false,
            lowerCase = false,
            digits = false,
            symbols = false
        ))
    }
}